### ISSUE TITLE
[module_variable_optional_attrs ] Terraform ignores changes in default section of the variable

### ISSUE THREAD

**Issue by Jasstkn** [hashicorp/terraform#30843]
### Terraform Version

```
1.1.8
```

### Terraform Configuration Files

```terraform
variable "prometheus_operator" {
  type = object({
    enabled     = bool,
    namespace   = string,
    version     = optional(string),
    values      = map(string),
    values_file = string,
  })
  default = {
    enabled   = true
    namespace = "monitoring"
    version   = "34.9.0" # <- version is changed
    values = {
    }
    values_file = "helm-values/prometheus-operator.yaml"
  }
}


resource "helm_release" "prometheus_operator" {
  count = var.prometheus_operator["enabled"] ? 1 : 0

  name             = var.prometheus_operator_release_name
  repository       = "https://prometheus-community.github.io/helm-charts"
  chart            = "kube-prometheus-stack"
  version          = var.prometheus_operator["version"]
 ...
}

```

### Expected Behavior
Terraform should detect changed version in the default section and reflect it in the plan

### Actual Behavior
No changes have been detected.

### Steps to Reproduce

1. `terraform init`
2. `terraform plan`

**Comment by alisdair** [comment]
Hi @Jasstkn, thanks for the report. Unfortunately I'm unable to reproduce this. Here's what I tried:

1. Create main.tf:
    ```terraform
    variable "config" {
      type = object({
        enabled = bool
        version = string
      })
      default = {
        enabled = true
        version = "5"
      }
    }

    resource "null_resource" "default" {
      count = var.config.enabled ? 1 : 0

      triggers = {
        config_version = var.config.version
      }
    }
    ```
2. `terraform init && terraform apply -auto-approve`
3. Update the variable's default value to change the version field to "6"
4. `terraform plan`

This results in a plan with changes, as expected.

It sounds like you're seeing this problem in a larger configuration. If you're able to reproduce it in isolation, please let us know the full series of steps so that we can investigate further.

**Comment by Jasstkn** [comment]
> Hi @Jasstkn, thanks for the report. Unfortunately I'm unable to reproduce this. Here's what I tried:
> 
> 1. Create main.tf:
>    ```terraform
>    variable "config" {
>      type = object({
>        enabled = bool
>        version = string
>      })
>      default = {
>        enabled = true
>        version = "5"
>      }
>    }
>    
>    resource "null_resource" "default" {
>      count = var.config.enabled ? 1 : 0
>    
>      triggers = {
>        config_version = var.config.version
>      }
>    }
>    ```
> 2. `terraform init && terraform apply -auto-approve`
> 3. Update the variable's default value to change the version field to "6"
> 4. `terraform plan`
> 
> This results in a plan with changes, as expected.
> 
> It sounds like you're seeing this problem in a larger configuration. If you're able to reproduce it in isolation, please let us know the full series of steps so that we can investigate further.

Hi @alisdair! Thank you for your prompt response. 
I was able to reproduce the issue using "from-scratch" approach. Looks like the issue occurs when I redefine the variable in the module block.
```
module "test" {
  source = "./module"

  prometheus_operator = {
    enabled   = true
    namespace = "default"

    values = tomap({})
  }
}
```
I create a sample structure that I use. First deploy it as it is, and then change default in the module's variable and see if terraform detect this change.
[demo.zip](https://github.com/hashicorp/terraform/files/8487858/demo.zip)

**Comment by alisdair** [comment]
> I expect values from default section to be merged with what I have in the module block but eventually something different happens.

Ah, I see the problem. This is not how the variable `default` value works: the `default` is only considered if the module's variable has no value assigned to it. Terraform does not merge the `default` with the passed value.

This is one of the design problems with the optional attributes/defaults experiment, which is why we are intending to redesign it for a future release. We advise against using this experiment, as it will be removed in a future Terraform version, and the replacement will likely not be directly compatible.

For the use case you describe, there are alternative ways to achieve the same goal. One is to break out the optional values into separate variables. For example, with the same module call, you can have an optional version with a default:

```terraform
variable "prometheus_operator" {
  type = object({
    enabled     = bool,
    namespace   = string,
    values      = map(string),
    values_file = string,
  })
  default = {
    enabled   = true
    namespace = "monitoring"
    values = {}
    values_file = "helm-values/prometheus-operator.yaml"
  }
}

variable "prometheus_operator_version" {
  type    = string
  default = "34.9.0" # this default is applied if the variable is not part of the module call
}

resource "helm_release" "prometheus_operator" {
  count = var.prometheus_operator["enabled"] ? 1 : 0

  name             = var.prometheus_operator_release_name
  repository       = "https://prometheus-community.github.io/helm-charts"
  chart            = "kube-prometheus-stack"
  version          = var.prometheus_operator_version # direct variable reference
 ...
}
```

I'm going to close this issue out for now, as it doesn't represent a bug in Terraform that we can fix. Hopefully we'll have a replacement for optional attributes in a release soon that addresses this issue by allowing you to specify individual default values within a complex variable type. Thanks again for the report!

**Comment by Jasstkn** [comment]
@alisdair hi. Thank you for clarifying this, I wasn't sure if it's expected behaviour or not but now I know. Looking forward for the next Terraform releases then 🎉

**Comment by korinne** [comment]
Hi all, we recently released the [Terraform v1.3 alpha](https://releases.hashicorp.com/terraform/1.3.0-alpha20220608), which includes the ability to mark object type attributes as optional and set default values.

I'm following up with issues that have provided feedback on the previous experiment, and wanted to invite you all to try the alpha and provide any feedback on this updated design. You can learn more/add comments[ here](https://discuss.hashicorp.com/t/request-for-feedback-optional-object-type-attributes-with-defaults-in-v1-3-alpha/40550). Thank you so much in advance, and hope you like the feature!
